WIP: try to enforce signing vs revocation

devrandom opened the pull request:
This fails in various places, ~~for example in `ln::chanmon_update_fail_tests::during_funding_monitor_fail`~~ because of persistence issues.

In particular, `EnforcingChannelKeys` would like to get some common state from `KeysInterface` when it's deserialized, but that's not available with the current persistence design.

devrandom committed:
Revocation enforcement in signer

We want to make sure that we don't sign revoked transactions.

Given that ChannelKeys are not singletons and revocation enforcement is stateful,
we need to store the revocation state in KeysInterface.

TheBlueMatt commented on lightning/src/util/enforcing_trait_impls.rs:
```diff
@@ -71,6 +96,19 @@ impl ChannelKeys for EnforcingChannelKeys {
+				if keys.per_commitment_point == self.inner.get_per_commitment_point(*revoked, secp_ctx) {
+					panic!("attempted to sign the latest revoked local commitment {}", self.inner.commitment_seed[0]);
+				} else {
+					panic!("can only sign the next two unrevoked commitment numbers, {} revoked={} point={}",
```
All the tests pass without this panic, and, indeed, I think you are correct in that it is just different objects getting serialized/read breaking things in the tests. It would be nice to enforce the other new assertions upstream even if we can't hit this one, but I think the solution is going to have to be some global static HashMap tracking revocation for a given commitment seed.

devrandom committed:
Introduce CommitmentTransaction, ChannelTransactionParameters

CommitmentTransaction maintains the per-commitment transaction fields needed to construct the associated bitcoin transactions (commitment, HTLC).  It replaces passing around of Bitcoin transactions.  The ChannelKeys API is modified accordingly.

By regenerating the transaction when implementing a validating external signer, this allows a higher level of assurance that all relevant aspects of the transactions were checked for policy violations.

ChannelTransactionParameters replaces passing around of individual per-channel fields that are needed to construct Bitcoin transactions.

Eliminate ChannelStaticData in favor of ChannelTransactionParameters.

Use counterparty txid instead of tx in channelmonitor update.

devrandom committed:
Unwrap unneeded macros

bmancini55 committed:
Add gossip_queries feature flag

Support for the gossip_queries feature flag (bits 6/7) is added to the
Features struct. This feature is available in the Init and Node
contexts. The gossip_queries feature is not fully implemented so this
feature is disabled when sent to peers in the Init message.

bmancini55 committed:
Add gossip_queries messages to wire decoding

To enable gossip_queries message decoding, this commit implements the
wire module's Encoding trait for each message type. It also adds these
messages to the wire module's Message enum and the read function to
enable decoding of a buffer.

bmancini55 committed:
Add send message events for gossip_queries

This change enables initiating gossip queries with a peer using the
SendMessageEvent enum. Specifically we add an event for sending
query_channel_range to discover the existance of channels and an event
for sending query_short_channel_ids to request routing gossip messages
for a set of channels. These events are handled inside the process_events
method of PeerManager which sends the serialized message to the peer.

bmancini55 committed:
Add gossip_queries methods to RoutingMessageHandler trait

Defines message handlers for gossip_queries messages in the RoutingMessageHandler
trait. The MessageSendEventsProvider supertrait is added to RoutingMessageHandler
so that the implementor can use SendMessageEvents to send messages to a
peer at the appropriate time.

The trait methods are stubbed in NetGraphMsgHandler which implements
RoutingMessageHandler and return a "not implemented" error.

bmancini55 committed:
Implement gossip_queries sync methods in NetGraphMsgHandler

To perform a sync of routing gossip messages with a peer requires a two
step process where we first initiate a channel range query to discover
channels in a block range. Next we request the routing gossip messages
for discovered channels.

This code implements logic in NetGraphMsgHandler for performing these two
tasks while taking into account the specification and variance in
implementation.

TheBlueMatt committed:
Cache fields instead of storing a ChannelKeys in ChannelMonitor

We only actually use two of the fields in ChannelKeys inside a
ChannelMonitor - the holder revocation_basepoint and the
derivation parameters. Both are relatively small, so there isn't
a lot of reason to hold a full copy of the ChannelKeys (with most
of the interaction with it being inside the OnchainTxHandler).

Further, this will avoid calling read on a `ChannelKeys` twice,
which is a somewhat strange API quirk.

TheBlueMatt committed:
Expose manual-update methods in NetworkGraph.

These functions were created but previously not exported, however
they are useful if we want to skip signature checking when accepting
routing messages (which we really should be doing in the routing
fuzzer).

TheBlueMatt committed:
[fuzz] Make router_target a bit easier for fuzzers to explore

It turns out (somewhat obviously) that expecting a fuzzer to
correctly build multiple signatures which verify against multiple
public keys in the same message was a bit too daunting, so we now
skip message signatures in routing messages.

We also take this opportunity to simplify the target itself somewhat,
avoiding reading public keys over and over and instead generating
routes to all the public keys that appeared in messages while running.

TheBlueMatt committed:
[router] Fix + test routing via next/last-hop hints only

We had code in the router to support sending a payment via a single
hop across channels exclusively provided by the next-/last-hop hints.
However, in updating the fuzzer, I noted that this case not only
didn't work, but paniced in some cases.

Here, we both fix the panic, as well as write a new test which
ensures we don't break support for such routing in the future.

TheBlueMatt committed:
[netgraph] Do not allow capacity_sats * 1000 to overflow-panic

In updating the router fuzzer, it discovered that a remote peer can
cause us to overflow while multiplying the channel capacity value.
Since the value should never exceed 21 million BTC, we just add a
check for that.

bmancini55 committed:
Refactor gossip_queries sync to be stateless

This commit simplifies the sync process for routing gossip messages. When
a sync is initiated, the process is handled statelessly by immediately
issuing SCID queries as channel range replies are received. This greatly
simplifies the state machine at the cost of fully validating and
conforming to the current spec.

TheBlueMatt committed:
Merge pull request #748 from TheBlueMatt/2020-11-router-fuzzer

Make router_target a bit easier for fuzzers to explore and fix two found bugs

TheBlueMatt committed:
Finally drop travis entirely.

It seems travis isn't even bothering to spawn our jobs anymore, and
Github CI has been pretty stable, so just drop Travis entirely.

TheBlueMatt committed:
Move UTXO-lookup into pub utility function from RoutingMsgHandler

This makes the public utility methods in `NetworkGraph` able to do
UTXO lookups ala `NetworkMsgHandler`'s `RoutingMessageHandler`
implementation, slightly simplifying the public interface.

We also take this opportunity to verify signatures before calling
out to UTXO lookups, under the "do actions in order of
cheapest-to-most-expensive to reduce DoS surface" principle.

TheBlueMatt committed:
Split `NetworkGraph` message handling fns into unsigned and signed

This takes the now-public `NetworkGraph` message handling functions
and splits them all into two methods - one which takes a required
Secp256k1 context and verifies signatures and one which takes only
the unsigned part of the message and does not take a Secp256k1
context.

This both clarifies the public API as well as simplifies it, all
without duplicating code.

Finally, this adds an assertion in the Router fuzzer to make sure
the constants used for message deserialization are correct.

TheBlueMatt committed:
Update auto-generated bindings with new NetworkGraph API

TheBlueMatt committed:
Bump version to 0.0.12

TheBlueMatt committed:
Merge pull request #758 from TheBlueMatt/2020-11-bindings-fix

Split `NetworkGraph` message handling fns into unsigned and signed (+update bindings)

TheBlueMatt committed:
Merge pull request #759 from TheBlueMatt/2020-11-0.0.12

Bump version to 0.0.12

TheBlueMatt committed:
Bump lightning-net-tokio crate version.

TheBlueMatt committed:
Merge pull request #760 from TheBlueMatt/2020-11-0.0.12

Bump lightning-net-tokio crate version.

TheBlueMatt committed:
Universally Require Writeable for ChannelKeys

It doesn't make sense to ever build a lightning node which doesn't
ever write ChannelMonitors to disk, so having a ChannelKeys object
which doesn't implement Writeable is nonsense.

Here we require Writeable for all ChannelKeys objects, simplifying
code generation for C bindings somewhat.

TheBlueMatt committed:
Add a new method `read_chan_signer` to `KeysInterface`

This adds a new method to the general cross-channel `KeysInterface`
which requires it to handle the deserialization of per-channel
signer objects. This allows the deserialization of per-channel
signers to have more context available, which, in the case of the
C bindings, includes the actual KeysInterface information itself.

TheBlueMatt committed:
Use Writeable for ChannelMonitor instead of a specific function.

There's no reason to have ChannelMonitor::write_for_disk instead of
just using the Writeable trait anymore. Previously, it was used to
differentiate with `write_for_watchtower`, but support for
watchtower-mode ChannelMonitors was never completed and the partial
bits were removed long ago.

This has the nice benefit of hitting the custom Writeable codepaths
in C bindings instead of trying to hit trait-generics paths.

TheBlueMatt committed:
Use `KeysInterface::read_chan_signer` for all channel keys deser

This drops any direct calls to a generic `ChannelKeys::read()` and
replaces it with the new `KeysInterface::read_chan_signer()`. Still,
under the hood all of our own `KeysInterface::read_chan_signer()`
implementations simply call out to a `Readable::read()` implemention.

TheBlueMatt committed:
Expose test_utils to fuzztarget (in addition to _test_utils feature)

bmancini55 committed:
Add genesis block hash to NetworkGraph

This changes adds the genesis block hash as a BlockHash to the
NetworkGraph struct. Making the NetworkGraph aware allows the message
handler to validate the chain_hash for received messages. This change
also adds the hash value to the Writeable and Readable methods.

devrandom committed:
Use TestKeysInterface in functional tests

This allows stateful validation in EnforcingChannelKeys

bmancini55 committed:
Pass gossip_queries messages to handler via ownership

This change modifies gossip_queries methods in RoutingMessageHandler to
move the message instead of passing a reference. This allows the message
handler to be more efficient by not requiring a full copy of SCIDs
passed in messages.

bmancini55 committed:
Pass Init message to sync_routing_table method

This commit modifies sync_routing_table in RoutingMessageHandler to
accept a reference to the Init message received by the peer. This allows
the method to use the Peer's features to drive the operations of the
gossip_queries routing table sync.

bmancini55 committed:
Change routing table sync to use gossip_queries

This commit changes outbound routing table sync to use gossip_queries
instead of the effectively deprecated initial_routing_sync feature.

This change removes setting of initial_routing_sync in our outbound Init
message. Instead we now call sync_routing_table after receiving an Init
message from a peer. If the peer supports gossip_queries and
should_request_full_sync returns true, we initiate a full gossip_queries
sync.

devrandom committed:
Fix fuzzing issue with revocation

bmancini55 committed:
Remove should_request_full_sync from RoutingMessageHandler

This method was used to set the initial_routing_sync flag when sending
an outbound Init message to a peer. Since we are now relying on
gossip_queries instead of initial_routing_sync, synchronization can be
fully encapsulate into RoutingMessageHandler via sync_routing_table.

This commit removes should_request_full_sync from the trait
RoutingMessageHandler. The implementation is still used in
NetGraphMsgHandler and has been converted into a private method instead
of a trait function.

bmancini55 committed:
Fix comment for sync_routing_table

Corrects the comment for sync_routing_table in RoutingMessageHandler to
be less prescriptive about the implementor's actions.

TheBlueMatt committed:
Merge pull request #736 from bmancini55/gossip_queries

Initiate gossip_queries

devrandom committed:
Use new get_commitment_transaction_number_obscure_factor utility function

TheBlueMatt committed:
Merge pull request #742 from lightning-signer/tx-phase2

Introduce CommitmentTransactionInfo

sr-gi committed:
Adds Copy to ChannelHandshakeConfig and UserConfig

This commits adds Copy to ChannelHandshakeConfig and UserConfig for consistency with the rest of config structs.

devrandom committed:
Fold sign_holder_commitment_htlc_transactions into sign_holder_commitment

Signing the commitment transaction is almost always followed by signing the attached HTLC transactions, so fold the signing operations into a single method.

devrandom committed:
Panic if signing fails in OnChainTx

Signatures in OnChainTx must not fail, or we stand to lose funds

TheBlueMatt committed:
Merge pull request #769 from sr-gi/copy-trait-config

Adds Copy to ChannelHandshakeConfig and UserConfig

TheBlueMatt committed:
Merge pull request #761 from TheBlueMatt/2020-10-chansigner-no-ser

Add a new method `read_chan_signer` to `KeysInterface`

jkczyz committed:
Add BOLT 3 test vector for CLTV tiebreaker

The test vector was added in https://github.com/lightningnetwork/lightning-rfc/pull/539.

Commit:
Add a GLOSSARY.md

Closes #766

Contributions by Devrandom <c1.devrandom@niftybox.net>

Commit:
Extend Getting Started with GLOSSARY/meeting minutes  in CONTRIBUTING.md

valentinewallace committed:
Update README with RL customization details

TheBlueMatt committed:
Merge pull request #771 from jkczyz/2021-01-bolt3-test-vectors

Add BOLT 3 test vector for CLTV tiebreaker

devrandom committed:
Let some tests disable revocation policy check

When simulating a bad actor that broadcasts a revoked tx, the policy check would otherwise panic.

devrandom committed:
Simplify handling of OnChainTx.holder_commitment

It is no longer optional since it is available at construction time.

TheBlueMatt committed:
Merge pull request #772 from ariard/2021-01-getting-project-context

Add GLOSSARY.md and more

TheBlueMatt committed:
Merge pull request #776 from valentinewallace/update-readme

Update readme

TheBlueMatt committed:
Merge pull request #770 from lightning-signer/holder-htlc-sigs

Fold sign_holder_commitment_htlc_transactions into sign_holder_commitment
